### PR TITLE
fix: bring back ability to automatically quote desired capabilities on win

### PR DESCRIPTION
**Pull request by root-intruder**
## Change list
- Bring back automatic quote escaping for desired capabilities command line arguments on windows
 
## Types of changes

What types of changes are you proposing/introducing to Java client?
_Put an `x` in the boxes that apply_

- [ ] No changes in production code.
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details
Escaped quotes are necessary for starting appium from the command line on windows systems (see http://appium.io/docs/en/writing-running-appium/default-capabilities-arg/). 

This capability to automatically quote on windows systems was completely removed with commit #1337. Since #1336 discovered a special case where automatic quote escaping on windows was not possible. That means, that windows users now need to manually escape the JSON such as:

```
DesiredCapabilities cap = new DesiredCapabilities();
cap.setCapability("\"platformName\"", "\"Android\"");

AppiumServiceBuilder builder = new AppiumServiceBuilder();
builder.withCapabilities(cap);
```
For a normal user this additional escape `"\"platformName\""` is cumbersome and not intuitive at all. It also makes it hard to write java code running on windows and unix/linux systems. 

From my understanding, the escaping should be done automatically by the java-client. However, this is a breaking change, thus I propose a switch to be able to enable quote escaping. If desired, I could also implement quote escaping as default behaviour with a switch to disable it. But I think this decision should be on your side.

Cheers!

### CODE REVIEW

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 1]
```diff
@@ -296,7 +315,48 @@ private void loadPathToMainScript() {
...
+            Set<? extends Map.Entry<String, ?>> entries = capabilitiesMap.entrySet();
+
+            for (Map.Entry<String, ?> entry : entries) {
+                Object value = entry.getValue();
+
+                if (value == null) {
+                    continue;
+                }
+
+                if (String.class.isAssignableFrom(value.getClass())) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:334]: value instanceof String should be enough

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 2]
```diff
@@ -296,7 +315,48 @@ private void loadPathToMainScript() {
...
+                    if (pathCapabilities.contains(entry.getKey())) {
+                        value = "\\\"" + String.valueOf(value).replace("\\", "/") + "\\\"";
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:336]: we could safely case value to String inside of this block

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 3]
```diff
@@ -296,7 +315,48 @@ private void loadPathToMainScript() {
...
+                    } else {
+                        value = "\\\"" + value + "\\\"";
+                    }
+                } else {
+                    value = String.valueOf(value);
+                }
+
+                String key = "\\\"" + entry.getKey() + "\\\"";
+                if (StringUtils.isBlank(result)) {
+                    result = key + ": " + value;
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: I usually prefer a ternary operator

Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: same above

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 2 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:336]: This was the code before commit #1337 so I decided to simply copy and paste this function to not break anyting.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 3 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: This was the code before commit #1337 so I decided to simply copy and paste this function to not break anyting.

Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: Should I still implement your suggestions?

Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: considering ternary operator, I disagree...usually makes code less readable.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 2 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: resolved in newest push

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 1 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: resolved in the newest push

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 4]
```diff
@@ -296,7 +315,49 @@ private void loadPathToMainScript() {
         this.appiumJS = findMainScript();
     }
 
+    private String parseCapabilitiesIfWindows() {
+        String result = StringUtils.EMPTY;
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:319]: Please replace it with StringBuilder

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 3 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:347]: I don't have a strong opinion on that, just a matter of taste

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 4 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: done in latest push

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 5]
```diff
@@ -296,7 +315,50 @@ private void loadPathToMainScript() {
...
+                    }
+                } else {
+                    value = String.valueOf(value);
+                }
+
+                String key = "\\\"" + entry.getKey() + "\\\"";
+                if (result.length() > 0) {
+                    result.append(", ");
+                }
+                result.append(key);
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:349]: this method returns self instance, so all calls could be changed like result.append().append()....

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 6]
```diff
@@ -296,7 +315,50 @@ private void loadPathToMainScript() {
...
+                        value = "\\\"" + valueString.replace("\\", "/") + "\\\"";
+                    } else {
+                        value = "\\\"" + valueString + "\\\"";
+                    }
+                } else {
+                    value = String.valueOf(value);
+                }
+
+                String key = "\\\"" + entry.getKey() + "\\\"";
+                if (result.length() > 0) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:346]: !result.isEmpty() could be more readable

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 5 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: you're right looks cleaner...pushed

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 7]
```diff
@@ -296,7 +315,48 @@ private void loadPathToMainScript() {
         this.appiumJS = findMainScript();
     }
 
+    private String parseCapabilitiesIfWindows() {
+        StringBuilder result = new StringBuilder();
+        final List<String> pathCapabilities = ImmutableList.of(AndroidMobileCapabilityType.KEYSTORE_PATH,
+                AndroidMobileCapabilityType.CHROMEDRIVER_EXECUTABLE, MobileCapabilityType.APP);
+
+        if (capabilities != null) {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:323]: this guard could be simplified to 
```
if (capabilities == null) {
  return "{}";
}
```

and moved to the beginning of the method

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 8]
```diff
@@ -296,7 +315,48 @@ private void loadPathToMainScript() {
         this.appiumJS = findMainScript();
     }
 
+    private String parseCapabilitiesIfWindows() {
+        StringBuilder result = new StringBuilder();
+        final List<String> pathCapabilities = ImmutableList.of(AndroidMobileCapabilityType.KEYSTORE_PATH,
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:320]: it makes sense to move this list to class-level constants

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 7 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:323]: I see no real advantage in your suggestion. The guard `if (capabilities == null) {` skips all methods code and directly jumps to the return at the end. Expressing it in the original reduces code duplication by not writing the return twice.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 8 above)
Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: done.

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java (hunk 7 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:323]: this improves readability. Longer code blocks are hard to scan through

Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: done

**Review by mykola-mokhnach — APPROVED** [review]
Thanks, I like it

File: src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java [hunk 9]
```diff
@@ -296,7 +317,46 @@ private void loadPathToMainScript() {
         this.appiumJS = findMainScript();
     }
 
+    private String parseCapabilitiesIfWindows() {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:320]: This method name does not do any check for the current platform, so I find it confusing it has `ifWindows` in its name

Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:320]: rename to ...for Windows?!

Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java:320]: how about `capabilitiesToQuotedCmdlineArg` ?

Comment (root-intruder) [reply src/main/java/io/appium/java_client/service/local/AppiumServiceBuilder.java]: sounds good aswell...changed

**Pull request merged**